### PR TITLE
KAFKA-10199: Handle task closure and recycling from state updater

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Within the tryCompleteRestore function of the thread, try to drain the removed tasks from state updater and handle accordingly: 1) for recycle, 2) for closure, 3) for update input partitions.
2. Catch up on some unit test coverage from previous PRs.
3. Some minor cleanups.
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 1]
```diff
@@ -298,13 +298,12 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
...
         );
 
-        final LinkedHashMap<TaskId, RuntimeException> taskCloseExceptions = new LinkedHashMap<>();
         final Map<TaskId, Set<TopicPartition>> activeTasksToCreate = new HashMap<>(activeTasks);
         final Map<TaskId, Set<TopicPartition>> standbyTasksToCreate = new HashMap<>(standbyTasks);
         final Map<Task, Set<TopicPartition>> tasksToRecycle = new HashMap<>();
         final Set<Task> tasksToCloseClean = new TreeSet<>(Comparator.comparing(Task::id));
 
-        tasks.purgePendingTasks(activeTasks.keySet(), standbyTasks.keySet());
+        tasks.purgePendingTasksToCreate(activeTasks.keySet(), standbyTasks.keySet());
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Do we really need to purge the pending tasks here?
Wouldn't it be clearer to clear the pending tasks to create? 
We would add the pending tasks anyways on line 318 and 319.
Maybe we can even first clear the pending tasks to create and then immediately call here:

```
tasks.addPendingActiveTasksToCreate(pendingTasksToCreate(activeTasksToCreate));
tasks.addPendingStandbyTasksToCreate(pendingTasksToCreate(standbyTasksToCreate));
```

In this way the pending tasks to create get removed from `activeTaskToCreate` and `standbyTaskToCreate` which is OK because the pending tasks have never been created, thus they will never be iterated over in `classifyTasks*()`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 2]
```diff
@@ -442,26 +440,25 @@ private void classifyTasksWithStateUpdater(final Map<TaskId, Set<TopicPartition>
         classifyRunningTasks(activeTasksToCreate, standbyTasksToCreate, tasksToRecycle, tasksToCloseClean);
         for (final Task task : stateUpdater.getTasks()) {
             final TaskId taskId = task.id();
+            final Set<TopicPartition> topicPartitions = activeTasksToCreate.get(taskId);
             if (activeTasksToCreate.containsKey(taskId)) {
                 if (task.isActive()) {
-                    final Set<TopicPartition> topicPartitions = activeTasksToCreate.get(taskId);
                     if (!task.inputPartitions().equals(topicPartitions)) {
-                        tasks.addPendingTaskThatNeedsInputPartitionsUpdate(taskId);
+                        tasks.addPendingTaskToUpdateInputPartitions(taskId, topicPartitions);
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:448]: Thank you for adding the topic partitions as a parameter.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TasksTest.java [hunk 3]
```diff
@@ -23,13 +23,16 @@
...
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
 import static org.apache.kafka.common.utils.Utils.mkSet;
 import static org.apache.kafka.test.StreamsTestUtils.TaskBuilder.standbyTask;
 import static org.apache.kafka.test.StreamsTestUtils.TaskBuilder.statefulTask;
 import static org.apache.kafka.test.StreamsTestUtils.TaskBuilder.statelessTask;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TasksTest.java]: nit: remove blank line

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreatorTest.java [hunk 4]
```diff
@@ -71,6 +72,8 @@ public class ActiveTaskCreatorTest {
     private StateDirectory stateDirectory;
     @Mock(type = MockType.NICE)
     private ChangelogReader changeLogReader;
+    @Mock(type = MockType.NICE)
+    private ProcessorStateManager stateManager;
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreatorTest.java]: This is never used.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java [hunk 5]
```diff
@@ -43,9 +43,7 @@
 class Tasks {
     private final Logger log;
 
-    // TODO: change type to `StreamTask`
     private final Map<TaskId, Task> activeTasksPerId = new TreeMap<>();
-    // TODO: change type to `StandbyTask`
     private final Map<TaskId, Task> standbyTasksPerId = new TreeMap<>();
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]: Why did you remove the ToDos?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java [hunk 6]
```diff
@@ -269,7 +271,6 @@ Collection<Task> tasks(final Collection<TaskId> taskIds) {
         return tasks;
     }
 
-    // TODO: change return type to `StreamTask`
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:272]: Why did you remove the ToDo?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 7]
```diff
@@ -641,9 +641,9 @@ boolean tryToCompleteRestoration(final long now, final java.util.function.Consum
                 }
             }
         } else {
-            for (final Task task : tasks.drainPendingTaskToRestore()) {
-                stateUpdater.add(task);
-            }
+            addTaskstoStateUpdater();
+
+            handleRemovedTasksFromStateUpdater();
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Could you add some unit tests for this code?

**Review by cadonna — COMMENTED** [review]
Thank you for the PR, @guozhangwang !

In `Tasks` on line 141 and line 163 why do we not verify that the task ID is contained in map and throw an `IllegalStateException` if not as we do for `standbyTasksPerId`? In the end those tasks are owned by the task.

Here is my other feedback!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: You're right! Will do so.

**Comment by guozhangwang** [comment]
> In Tasks on line 141 and line 163 why do we not verify that the task ID is contained in map and throw an IllegalStateException if not as we do for standbyTasksPerId? In the end those tasks are owned by the task.

Following your other comment, I've removed these two lines.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreatorTest.java (hunk 4 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreatorTest.java]: Sorry, forget to remove it; it's for another work to address comments for the test coverage.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]: I tried to convert the types to StreamTask/StandbyTask, but realized that the blocker from doing so is because we have the stub class `StateMachineTask` in other unit tests which are not either. I then looked into if there's a better way to remove that stub class and instead just use a mock StreamTask/StandbyTask but that's actually not very convienent. So I figured unless we'd eventually remove `StateMachineTask` we'd have to keep as is. 

In a second thought, I think we'd probably want to remove it eventually, so I'd add those TODOs back.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 8]
```diff
@@ -656,6 +657,71 @@ boolean tryToCompleteRestoration(final long now, final java.util.function.Consum
...
+
+                    if (task.state() != State.CLOSED) {
+                        tasksToCloseDirty.add(task);
+                    }
+
+                    taskExceptions.putIfAbsent(task.id(), e);
+                }
+            } else if ((inputPartitions = tasks.removePendingTaskToUpdateInputPartitions(task.id())) != null) {
+                task.updateInputPartitions(inputPartitions, topologyMetadata.nodeToSourceTopics(task.id()));
+                stateUpdater.add(task);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:715]: Found this issue when adding unit test: after update the input partitions, need to add it back to state updater.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 9]
```diff
@@ -656,6 +657,71 @@ boolean tryToCompleteRestoration(final long now, final java.util.function.Consum
...
+
+                    if (task.state() != State.CLOSED) {
+                        tasksToCloseDirty.add(task);
+                    }
+
+                    taskExceptions.putIfAbsent(taskId, e);
+                }
+            } else if (tasks.removePendingTaskToClose(task.id())) {
+                try {
+                    task.closeClean();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:698]: Should not remove from the `tasks` since it's not in it, hence we could not call `closeClean`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 10]
```diff
@@ -1076,59 +1135,51 @@ private Collection<Task> tryCloseCleanActiveTasks(final Collection<Task> activeT
...
-                final RuntimeException exception = completeTaskCloseClean(task);
-                if (exception != null) {
-                    firstException.compareAndSet(null, exception);
-                }
-            } catch (final StreamsException e) {
-                log.error("Exception caught while clean-closing task " + task.id(), e);
-                e.setTaskId(task.id());
-                firstException.compareAndSet(null, e);
-                tasksToCloseDirty.add(task);
+                closeTaskClean(task);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:1178]: This is the minor cleanup for exception wrapping / handling.

**Review by guozhangwang — COMMENTED** [review]
Thanks @cadonna for the reviews! I've addressed your comments and added unit tests. It's good for a final look now.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 11]
```diff
@@ -316,20 +313,22 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
             classifyTasksWithStateUpdater(activeTasksToCreate, standbyTasksToCreate, tasksToRecycle, tasksToCloseClean);
         }
 
-        tasks.addPendingActiveTasks(pendingTasksToCreate(activeTasksToCreate));
-        tasks.addPendingStandbyTasks(pendingTasksToCreate(standbyTasksToCreate));
+        tasks.purgePendingTasksToCreate();
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: nit: `purgePendingTasksToCreate()` -> `clearPendingTasksToCreate()`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 12]
```diff
@@ -316,20 +313,22 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
...
+        tasks.addPendingActiveTasksToCreate(pendingTasksToCreate(activeTasksToCreate));
+        tasks.addPendingStandbyTasksToCreate(pendingTasksToCreate(standbyTasksToCreate));
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: I was actually proposing to move this block before line 306.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java [hunk 13]
```diff
@@ -43,35 +43,48 @@
...
 
-    void purgePendingTasksToCreate(final Set<TaskId> assignedActiveTasks, final Set<TaskId> assignedStandbyTasks) {
-        pendingActiveTasksToCreate.keySet().retainAll(assignedActiveTasks);
-        pendingStandbyTasksToCreate.keySet().retainAll(assignedStandbyTasks);
+    void purgePendingTasksToCreate() {
+        pendingActiveTasksToCreate.clear();
+        pendingStandbyTasksToCreate.clear();
+    }
+
+    Map<TaskId, Set<TopicPartition>> drainPendingActiveTasksForTopologies(final Set<String> currentTopologies) {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:72]: Could you provide unit tests for this method?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java [hunk 14]
```diff
@@ -43,35 +43,48 @@
...
+        final Map<TaskId, Set<TopicPartition>> pendingActiveTasksForTopologies =
+            filterMap(pendingActiveTasksToCreate, t -> currentTopologies.contains(t.getKey().topologyName()));
+
+        pendingActiveTasksToCreate.keySet().removeAll(pendingActiveTasksForTopologies.keySet());
+
+        return pendingActiveTasksForTopologies;
+    }
+
+    Map<TaskId, Set<TopicPartition>> pendingStandbyTasksForTopologies(final Set<String> currentTopologies) {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]: Could you provide unit tests for this method?

I would also rename this method to `drainPendingStandbyTasksForTopologies` to be consistent with the previous method.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 15]
```diff
@@ -296,6 +296,75 @@ public void shouldAddTasksToStateUpdater() {
...
+        when(task02.inputPartitions()).thenReturn(taskId02Partitions);
+        when(task03.inputPartitions()).thenReturn(taskId03Partitions);
+        when(task00.isActive()).thenReturn(true);
+        when(task01.isActive()).thenReturn(false);
+        when(task02.isActive()).thenReturn(true);
+        when(task03.isActive()).thenReturn(true);
+        when(task00.state()).thenReturn(State.RESTORING);
+        when(task01.state()).thenReturn(State.RUNNING);
+        when(task02.state()).thenReturn(State.RESTORING);
+        when(task03.state()).thenReturn(State.RESTORING);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: We should extend the mock task builder in `StreamsTestUtils` to allow the specification of the input partitions. This setup code would get much simpler with that. I can extends the utils once this PR is merged.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 16]
```diff
@@ -296,6 +296,75 @@ public void shouldAddTasksToStateUpdater() {
...
+        expect(activeTaskCreator.createActiveTaskFromStandby(eq(task01), eq(taskId01Partitions), eq(consumer)))
+            .andStubReturn(task01Converted);
+        activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(anyObject());
+        expectLastCall().times(2);
+        expect(standbyTaskCreator.createStandbyTaskFromActive(eq(task00), eq(taskId00Partitions)))
+            .andStubReturn(task00Converted);
+        expect(consumer.assignment()).andReturn(emptySet()).anyTimes();
+        consumer.resume(anyObject());
+        expectLastCall().anyTimes();
+        replay(activeTaskCreator, standbyTaskCreator, topologyBuilder, consumer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Why are you mixing Mockito with EasyMock? Since this is new code you could just use Mockito.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 17]
```diff
@@ -296,6 +296,75 @@ public void shouldAddTasksToStateUpdater() {
...
+    public void shouldHandleRemovedTasksFromStateUpdater() {
+        // tasks to recycle
+        final StreamTask task00 = mock(StreamTask.class);
+        final StandbyTask task01 = mock(StandbyTask.class);
+        final StandbyTask task00Converted = mock(StandbyTask.class);
+        final StreamTask task01Converted = mock(StreamTask.class);
+        // task to close
+        final StreamTask task02 = mock(StreamTask.class);
+        // task to update inputs
+        final StreamTask task03 = mock(StreamTask.class);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I would add a test for each of these task lifecycle actions. This makes the tests easier readable.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 18]
```diff
@@ -296,6 +296,75 @@ public void shouldAddTasksToStateUpdater() {
         Mockito.verify(stateUpdater).add(task01);
     }
 
+    @Test
+    public void shouldHandleRemovedTasksFromStateUpdater() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:387]: The exceptional cases are missing like `newTask.initializeIfNeeded();` throws an exception or closing a tasks throws an exception.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 19]
```diff
@@ -641,9 +642,9 @@ boolean tryToCompleteRestoration(final long now, final java.util.function.Consum
                 }
             }
         } else {
-            for (final Task task : tasks.drainPendingTaskToRestore()) {
-                stateUpdater.add(task);
-            }
+            addTaskstoStateUpdater();
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: `addTaskstoStateUpdater()` -> ` addTasksToStateUpdater()`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 18 above)
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:387]: We also need a test to verify that new tasks are initialized and added to the state updater.

**Review by cadonna — COMMENTED** [review]
@guozhangwang Thanks for the updates!

Here my feedback!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 15 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I opened https://github.com/apache/kafka/pull/12510 for extending the mock task builder. If you think it makes sense you could approve and merge https://github.com/apache/kafka/pull/12510, rebase this PR and use the builder here.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 18 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:387]: I just saw that `shouldAddTasksToStateUpdater()` is such a test. The verification for calling `initializeIfNeeded()` is missing, though.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 9 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:698]: Don't we need to suspend the task before, we cleanly close it?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 20]
```diff
@@ -296,6 +296,75 @@ public void shouldAddTasksToStateUpdater() {
...
+            UUID.randomUUID(),
+            "taskManagerTest",
+            activeTaskCreator,
+            standbyTaskCreator,
+            topologyMetadata,
+            adminClient,
+            stateDirectory,
+            stateUpdater
+        );
+        taskManager.setMainConsumer(consumer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]:
```suggestion
        final TaskManager taskManager = setUpTaskManager(ProcessingMode.AT_LEAST_ONCE, true);
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 16 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Now I see what the issue is. Some global mocks like the main consumer are specified using EasyMock which makes using Mockito hard without changing all other test methods.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:698]: You're right! Added and also updated unit tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: ack.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 16 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Yes.. I tried to use Mockito for all and soon realized it would incur a much larger scope.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java:72]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java (hunk 14 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]: Ack.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 15 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Ack. Rebased on your utils and break it down to multiple tests.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 18 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:387]: This test did verify that `initializeIfNeeded` is called, and the tasks are added back to the state updater.

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:387]: When adding tests I realized we do not need to handle exceptions by closing tasks dirty and then re-thrown, but could simply re-throw directly and delegate to the StreamThread to handle the thrown exception. Will updated accordingly.

**Comment by guozhangwang** [comment]
Merged to trunk, @cadonna please feel free to leave more comments, and I will address in https://github.com/apache/kafka/pull/12519
